domainlab/algos/trainers/fbopt.py

```diff
@@ -0,0 +1,95 @@
+"""
+update hyper-parameters during training
+"""
+import copy
+from domainlab.utils.logger import Logger
+
+
+class HyperSchedulerFeedback():
+    """
+    design $\\mu$$ sequence based on state of penalized loss
+    """
+    def __init__(self, trainer, **kwargs):
+        """
+        kwargs is a dictionary with key the hyper-parameter name and its value
+        """
+        self.trainer = trainer
+        self.mmu = kwargs
+        self.mmu = {key: 0.0 for key, val in self.mmu.items()}
+        self.ploss_old_theta_old_mu = None
+        self.ploss_old_theta_new_mu = None
+        self.ploss_new_theta_old_mu = None
+        self.ploss_new_theta_new_mu = None
+        self.delta_mu = 0.01   # FIXME
+        self.dict_theta = None
+        self.budget_mu_per_step = 5  # FIXME
+        self.budget_theta_update_per_mu = 5  # np.infty
+
+    def search_mu(self, dict_theta):
+        """
+        start from parameter dict_theta,
+        enlarge mmu to see if the criteria is met
+        """
+        self.dict_theta = dict_theta
+        mmu = None
+        for miter in range(self.budget_mu_per_step):
+            # FIXME: the same mu is tried two times since miter=0
+            mmu = self.dict_addition(self.mmu, miter * self.delta_mu)
```

Comment:
miter always start from 0, but let miter=0 only make sense for the 1st iteration